testgrid/cmd/updater/main.go

```diff
@@ -531,6 +527,141 @@ func ValidateName(name string) map[string]string {
+func readJson(ctx context.Context, obj *storage.ObjectHandle, i interface{}) error {
+	reader, err := obj.NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("open: %v", err)
+	}
+	if err = json.NewDecoder(reader).Decode(i); err != nil {
+		return fmt.Errorf("decode: %v", err)
+	}
+	return nil
+}
+
+// readStarted parses the build's started metadata.
+func readStarted(ctx context.Context, build Build) (*Started, error) {
+	uri := build.Prefix + "started.json"
+	var started Started
+	if err := readJson(ctx, build.Bucket.Object(uri), &started); err != nil {
+		return nil, fmt.Errorf("read %s: %v", uri, err)
+	}
+	return &started, nil
+}
+
+// readFinished parses the build's finished metadata.
+func readFinished(ctx context.Context, build Build) (*Finished, error) {
+	uri := build.Prefix + "finished.json"
+	var finished Finished
+	if err := readJson(ctx, build.Bucket.Object(uri), &finished); err != nil {
+		return nil, fmt.Errorf("read %s: %v", uri, err)
+	}
+	return &finished, nil
+}
+
+// listArtifacts writes the object name of everything in the build's artifacts dir to the output chan.
+func listArtifacts(ctx context.Context, build Build, output chan string) error {
+	pref := build.Prefix + "artifacts/"
+	objs := build.Bucket.Objects(ctx, &storage.Query{Prefix: pref})
+	for {
+		obj, err := objs.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed to list %s: %v", pref, err)
+		}
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("interrupted listing %s", pref)
+		case output <- obj.Name: // Sent to chan
+		}
+	}
+	return nil
+}
+
+// readSuites parses the <testsuite> or <testsuites> object in obj
+func readSuites(ctx context.Context, obj *storage.ObjectHandle) (*junit.Suites, error) {
+	reader, err := obj.NewReader(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("open: %v", err)
+	}
+
+	buf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("read: %v", err)
+	}
+
+	suites, err := junit.Parse(buf)
+	if err != nil {
+		return nil, fmt.Errorf("parse: %v", err)
+	}
+	return &suites, nil
+}
+
+type suitesMeta struct {
+	suites   junit.Suites      // suites data extracted from file contents
+	metadata map[string]string // metadata extracted from path name
+}
+
+// listSuites takes a list of artifact names, parses those representing junit suites, writing the result to the output channel.
+//
+// Note that junit suites are parsed in parallel, so there are no guarantees about suites ordering.
+func listSuites(ctx context.Context, build Build, artifacts chan string, output chan suitesMeta) error {
+
+	var wg sync.WaitGroup
+	ec := make(chan error)
+	defer close(ec)
+	// since this function spawns go-routines, we need to be able to cancel those go-routines.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for art := range artifacts {
+		meta := validateName(art) // do we think this path holds a junit suites file?
```

Comment:
Could you rename the function to `isValidJunitName()` and get rid of the comments?

Reply:
Done